File: agapay-mobile/test/support/predictions.dart
Map<String, dynamic> predictionFixture({
  String status = 'not_trained',
  String source = 'none',
  DateTime? now,
}) {
  final at = (now ?? DateTime.now()).toUtc();
  return {
    'schema_version': '1.0',
    'station_id': 'STATION_001',
    'status': status,
    'reason': source == 'simulated'
        ? 'Sample response for interface testing.'
        : 'A trained forecast model is not configured.',
    'source': source,
    'advisory_only': true,
    'checked_at': at.toIso8601String(),
    'generated_at': status == 'available' ? at.toIso8601String() : null,
    'valid_until': status == 'available'
        ? at.add(const Duration(seconds: 60)).toIso8601String()
        : null,
    'input_last_recorded_at': at.toIso8601String(),
    'input_window_start': at
        .subtract(const Duration(minutes: 55))
        .toIso8601String(),
    'model_version': source == 'model' ? 'test-only' : null,
    'preview_allowed': true,
    'input_status': 'usable',
    'sample_count': 12,
    'valid_sample_count': 12,
    'timestamp_basis': source == 'simulated' ? 'synthetic' : 'server_received',
    'rainfall_interval_seconds': null,
    'forecasts': [
      for (var i = 0; i < 3; i++)
        {
          'horizon_minutes': (i + 1) * 30,
          'water_depth_cm': status == 'available' ? 96.0 + 3 * i : null,
          'target_at': status == 'available'
              ? at.add(Duration(minutes: (i + 1) * 30)).toIso8601String()
              : null,
        },
    ],
  };
}

File: agapay-web/src/lib/predictions.ts
export const forecastLabels = {
  not_trained: "Model not trained", insufficient_data: "Insufficient history", stale_data: "Readings are stale",
  invalid_data: "Invalid readings", available: "Forecast available", service_error: "Service unavailable",
} as const;
export type ForecastStatus = keyof typeof forecastLabels;
export type Prediction = {
  schema_version: "1.0"; station_id: string; status: ForecastStatus; reason: string;
  source: "none" | "model" | "simulated"; advisory_only: true;
  checked_at: string; generated_at: string | null; valid_until: string | null; model_version: string | null;
  input_status: "usable" | "insufficient_data" | "stale_data" | "invalid_data";
  input_window_start: string | null; input_last_recorded_at: string | null;
  sample_count: number; valid_sample_count: number; preview_allowed: boolean;
  timestamp_basis: "server_received" | "device_measured" | "synthetic";
  rainfall_interval_seconds: number | null;
  forecasts: { horizon_minutes: 30 | 60 | 90; target_at: string | null; water_depth_cm: number | null }[];
};

// Fail closed when the adapter/API contract changes or returns malformed values.
export function parsePrediction(value: unknown): Prediction {
  const p = value as Prediction;
  const date = (v: unknown) => typeof v === "string" && Number.isFinite(Date.parse(v));
  if (!p || p.schema_version !== "1.0" || p.advisory_only !== true || !Object.hasOwn(forecastLabels, p.status)
    || !["none", "model", "simulated"].includes(p.source) || typeof p.station_id !== "string"
    || typeof p.reason !== "string" || !date(p.checked_at) || typeof p.preview_allowed !== "boolean"
    || !Array.isArray(p.forecasts) || p.forecasts.length !== 3
    || p.forecasts.some((v, i) => !v || v.horizon_minutes !== [30, 60, 90][i])) throw new Error("The forecast response could not be validated.");
  if (p.status === "available") {
    if (p.source === "none" || !date(p.generated_at) || !date(p.input_last_recorded_at) || !date(p.valid_until)
      || Date.parse(p.valid_until!) <= Date.parse(p.checked_at)
      || (p.source === "model" && !p.model_version)
      || p.forecasts.some((v) => typeof v.water_depth_cm !== "number" || !Number.isFinite(v.water_depth_cm) || v.water_depth_cm < 0 || v.water_depth_cm > 1000 || !date(v.target_at))) throw new Error("The forecast response could not be validated.");
  } else if (p.forecasts.some((v) => v.water_depth_cm !== null || v.target_at !== null)) throw new Error("Unavailable forecasts contained unexpected values.");
  if (p.source === "simulated" && p.model_version !== null) throw new Error("A sample preview cannot claim a trained model.");
  return p;
}

export function forecastTime(value: string | null) {
  return value ? new Date(value).toLocaleString("en-PH", { month: "short", day: "numeric", hour: "2-digit", minute: "2-digit", timeZone: "Asia/Manila" }) : "—";
}
